PP-856: Implement AppVeyor CI for Windows

hirenvadalia opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Issue-ID
<!--- Replace XXXX below (2 places) with the actual JIRA id -->
* **[PP-856](https://pbspro.atlassian.net/browse/PP-856)**

#### Problem description
* No CI for windows

#### Cause / Analysis
* Time crunch while open sourcing PBS Pro 

#### Solution description
* Implemented AppVeyor CI

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [ ] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [ ] I have attached **test logs to the Jira ticket** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

hirenvadalia commented:
Please have a look at [here](https://ci.appveyor.com/project/hirenvadalia/pbspro/build/%23220) to see results on AppVeyor

arungrover commented on src/lib/Libutil/pbs_aes_encrypt.c:
```diff
@@ -33,27 +33,28 @@ extern unsigned char pbs_aes_iv[];
 pbs_encrypt_data(char *uncrypted, int *credtype, size_t len, char **crypted, size_t *outlen)
 {
-        EVP_CIPHER_CTX ctx;
+        EVP_CIPHER_CTX *ctx = EVP_CIPHER_CTX_new();
```
why did we move it from stack to heap?

hirenvadalia replied:
Hey @arungrover,
First thanks a lot reviewing this. 😄 
This change is due to there is change in latest version of openssl about how we can access internal structure of API.
Please have a look at [here](https://github.com/openssl/openssl/issues/962#issuecomment-208792020) for more information.
Please let me know if you need more information.

arungrover commented on src/lib/Libutil/pbs_aes_encrypt.c:
```diff
@@ -33,27 +33,28 @@ extern unsigned char pbs_aes_iv[];
                 return -1;
 
-        if (EVP_EncryptUpdate(&ctx, cblk, &plen, uncrypted, len) == 0)
+        if (EVP_EncryptUpdate(ctx, cblk, &plen, uncrypted, len) == 0)
```
Don't you want to call EVP_CIPHER_CTX_free before returning?

hirenvadalia replied:
Yes we do, will update code. Thank for catching!

arungrover commented on src/lib/Libutil/pbs_aes_encrypt.c:
```diff
@@ -33,27 +33,28 @@ extern unsigned char pbs_aes_iv[];
                 return -1;
 
-        EVP_CIPHER_CTX_cleanup(&ctx);
+        EVP_CIPHER_CTX_cleanup(ctx);
```
These two lines have a mix of tabs and spaces

hirenvadalia replied:
will update.

arungrover commented on src/lib/Libutil/pbs_aes_encrypt.c:
```diff
@@ -90,27 +91,28 @@ pbs_encrypt_pwd(char *str, int *credtype, char **credbuf, size_t *credlen)
                 return -1;
 
-        if (EVP_DecryptUpdate(&ctx, cblk, &plen, crypted, len) == 0)
+        if (EVP_DecryptUpdate(ctx, cblk, &plen, crypted, len) == 0)
```
Don't you want to free cblk and ctx before returning?

hirenvadalia replied:
Yes we do, will update code. Thank for catching!

arungrover commented on src/lib/Libutil/pbs_aes_encrypt.c:
```diff
@@ -90,27 +91,28 @@ pbs_encrypt_pwd(char *str, int *credtype, char **credbuf, size_t *credlen)
 
-        EVP_CIPHER_CTX_cleanup(&ctx);
+        EVP_CIPHER_CTX_cleanup(ctx);
+		EVP_CIPHER_CTX_free(ctx);
```
indentation

hirenvadalia replied:
oops... will update.

arungrover commented on .appveyor/build_libedit.bat:
```diff
@@ -0,0 +1,59 @@
+
+cd "%BUILDDIR%"
+
+if exist "%BINARIESDIR%\libedit" (
```
If it is a CI platform, what are the chances that libedit, postgres and other such third party will already be present on the machine? If they are then why do we error out?

hirenvadalia replied:
On CI there are no chances that any of this directory is exists. I have added this check for specifically developers, because I don't see a point of building all PBS deps from start if you have already built.
However we should not error out if directory exists, it should be exit 0, I will change code.

arungrover replied:
This is an interesting use case. It seems this change addresses more than just appveyor CI builds.
Are there special instructions on how developers can build PBS on their windows machine?
Can you post test logs of such a compilation?

hirenvadalia replied:
As such there are no special instructions for developers to build PBS on windows using these scripts, you just have to run 3 steps mentioned in .appveyor.yml file but however these scripts to works you need some tools to be installed in system (like MinGW, SVN, Perl, Cmake etc... for more info check set_paths.bat) and based on path where you installed this tool modify set_paths.bat file.

arungrover commented on .appveyor/build_python.bat:
```diff
@@ -0,0 +1,229 @@
+        "%MSYSDIR%\bin\bash" --login -i -c "cd $BUILDDIR_M/cpython-2.7.13 && tar -xf $BUILDDIR_M/python_externals.tar.gz"
+    )
+) else (
+	call "%BUILDDIR%\cpython-2.7.13\PCbuild\get_externals.bat"
```
do we need to call this while building 32 bit Python?

hirenvadalia replied:
No, because PC\VS9.0\build.bat with call it if you pass '-e' argument to build.bat, I am calling here specifically because we have to delete one file (externals\openssl-1.0.2j\ms\nt64.mak) from openssl directory as there is bug (https://bugs.python.org/issue30368) in Python 2.7.13 source code, without deleting this file it won't build ssl module in 64 bit.

arungrover commented on .appveyor/build_python.bat:
```diff
@@ -0,0 +1,229 @@
+cd "%BUILDDIR%\cpython-2.7.13"
+
+REM Restore externals directory if python_externals.tar.gz exists
+if exist "%BUILDDIR%\python_externals.tar.gz" (
```
why do we check for it again?

hirenvadalia replied:
We previously we checked it while compiling 32 bit Python this is while compiling 64 bit Python.
And just FYI, once we complete 32 bit compilation we remove "%BUILDDIR%\cpython-2.7.13" directory and extract again from %BUILDDIR%\cpython-2.7.13.zip to clean build of 64 bit python.

arungrover commented on .appveyor/build_python.bat:
```diff
@@ -0,0 +1,229 @@
+)
+
+REM Workaround for python2713.chm
+mkdir Doc\build\htmlhelp
```
Have we not done this already?

hirenvadalia replied:
Yes, but that was while compiling 32 bit Python, this is while compiling 64 bit Python.

arungrover commented on src/lib/Libutil/pbs_aes_encrypt.c:
```diff
@@ -33,27 +33,45 @@ extern unsigned char pbs_aes_iv[];
+        if (!cblk) {
+                EVP_CIPHER_CTX_cleanup(ctx);
+                EVP_CIPHER_CTX_free(ctx);
+                cblk = NULL;
```
this assignment is not harmful but it is not necessary either

hirenvadalia replied:
Yup, will remove it.

hirenvadalia commented on .appveyor/appveyor_helper.bat:
```diff
@@ -0,0 +1,40 @@
+@echo off
+setlocal enableDelayedExpansion
```
Nice question 👍 
I used enableDelayedExpansion for following two reason:
1. To expand variables at execution time rather than parse time (see [here](https://ss64.com/nt/delayedexpansion.html) for more information
2. Restore original environment at end of batch file (see [here](https://stackoverflow.com/a/6680005) or ```setlocal /?``` in cmd for more information)

hirenvadalia commented on .appveyor/appveyor_helper.bat:
```diff
@@ -0,0 +1,40 @@
+setlocal enableDelayedExpansion
+
+rem for /f %%f in ('wmic cpu get NumberOfLogicalProcessors /value ^| find "="') do set %%f
+rem set /a "ncpus=%NumberOfLogicalProcessors%*%NUMBER_OF_PROCESSORS%"
```
No this are not comment, actually while development I was trying to find out actual number for core available in system. Anyways this is not needed anymore I will remove them as we need only number of processors only.

hirenvadalia commented on .appveyor/appveyor_helper.bat:
```diff
@@ -0,0 +1,40 @@
+set hasmore=1
+1>nul 2>nul del *.out *.passed
+for /f "usebackq" %%A in (`dir /on /b .appveyor ^| findstr /B /R "^build_.*\.bat$"`) do (
+    if !started! lss %NUMBER_OF_PROCESSORS% (
```
(If you want I can put comments in code also)
Actually this script is just helper script to build PBS deps in parallel to reduce time of building all PBS deps.
Now on to your question regarding how this script works:
1. it will list all build_*.bat file in .appveyor directory
2. one by one it will execute each batch file in background while capturing output and error or batchfile in \<batch file name>.bat.out file in current directory
3. while executing batch file it will limit number of started background process to number of processors available in system.
4. once it started background process reached limit it goes to wait loop where it will wait for atleast one background process to complete by looking at exclusive lock on \<batch file name>.bat.out created by start of process and that lock will be released as soon as process ends
5. as soon as one processor ends it checks for \<batch file name>.bat.passed file if it exists that means background process completed successfully else some error occurred. Now in case of error this script will show contents of \<batch file name>.bat.out to console else it will just put one line saying \<batch file name>.bat finished.
6. Now since one process if finished we have one room to start new process so it will go back to main loop to start new background process for building next PBS deps and main loop again goes to wait loop and this goes on till end of all batch file in .appveyor directory

Hope you understand how this script works! Please do not hesitate to ask question if you don't 😄

arungrover replied:
Major part of confusion was your comments in previous commits which mentioned "wmic" being used and then the code was just using NUMBER_OF_PROCESSORS. It's good that you got rid of it.
This description is helpful. Please add to point number 5 on what happens to this batch script when one of the third party compilation fails.

hirenvadalia replied:
Oh! Sorry, I didn't know that. Sure I will modify point 5. Thanks.

arungrover commented on .appveyor.yml:
```diff
@@ -0,0 +1,13 @@
+  - Release
+clone_depth: 5
+build_script:
+  - cmd: call .appveyor\appveyor_helper.bat
```
So what happens when appveyor_helper.bat exits with an error? would it still continue to build PBS?

hirenvadalia replied:
No (as Travis) as soon as this steps fails build will stops immediately with error.

hirenvadalia replied:
@arungrover We just got one build fail where we can see that what happens if any of PBS deps fails... see [here](https://ci.appveyor.com/project/hirenvadalia/pbspro/build/%23228)

hirenvadalia commented:
@arungrover I have addressed your comments + some enhancements, please have a look in latest commit and let me know your thoughts.

kjakkali commented on .appveyor/build_swig.bat:
```diff
@@ -0,0 +1,94 @@
+        exit /b 1
+    )
+)
+if not exist "%BUILDDIR%\cccl-1.0.zip" (
```
Can you please add a comment about why we need cccl library?

kjakkali commented on .appveyor/set_paths.bat:
```diff
@@ -0,0 +1,17 @@
+set PERL_BIN=perl
+set CMAKE_BIN=cmake
+set BUILDDIR=%CD%
+set BINARIESDIR=%BUILDDIR%\binaries
```
In one of the above comment you mention about SVN dependency. I don't see any reference here.

kjakkali commented on win_configure/projects.VS2008/pbs_tclsh.vcproj:
```diff
@@ -45,7 +45,7 @@
 				Optimization="2"
 				InlineFunctionExpansion="1"
-				AdditionalIncludeDirectories="&quot;c:\Program Files\tcl\include&quot;;..\..\src\include;..\include;c:\tcl\include"
+				AdditionalIncludeDirectories="..\..\..\binaries\tcltk\include;..\..\src\include;..\include;c:\tcl\include"
```
Do we still need c:\tcl\include?

kjakkali commented on win_configure/projects.VS2008/qstat.vcproj:
```diff
@@ -44,7 +44,7 @@
 				Name="VCCLCompilerTool"
 				Optimization="0"
-				AdditionalIncludeDirectories="..\..\src\include;..\include;&quot;c:\Program Files\Tcl\Include&quot;;\tcl\include"
+				AdditionalIncludeDirectories="..\..\src\include;..\include;..\..\..\binaries\tcltk\include;\tcl\include"
```
Do we need '\tcl\include'?

kjakkali commented on win_configure/projects.VS2008/svr_periodic_hook.vcproj:
```diff
@@ -168,11 +168,11 @@
 				LinkIncremental="1"
 				SuppressStartupBanner="true"
-				AdditionalLibraryDirectories="..\..\..\win_build\src\server\Release;&quot;c:\Program Files\LM-X\lib\win32_x86&quot;;&quot;c:\Program Files\Openssl\lib&quot;;&quot;c:\Program Files\Python-2.7.11\PC\VS9.0\Release&quot;;&quot;c:\Program Files\libical\lib&quot;;&quot;C:\Program Files\pgsql\lib&quot;"
+				AdditionalLibraryDirectories="..\..\..\win_build\src\server\Release;..\..\..\binaries\openssl\lib;;..\..\..\binaries\python\libs;..\..\..\binaries\libical\lib;;..\..\..\binaries\pgsql\lib;"
```
Replace ;; with ;

kjakkali commented on win_configure/projects.VS2008/pbs_tclsh.vcproj:
```diff
@@ -165,11 +165,11 @@
 				LinkIncremental="1"
 				SuppressStartupBanner="true"
-				AdditionalLibraryDirectories="&quot;c:\Program Files\Tcl\lib&quot;;c:\tcl\lib"
+				AdditionalLibraryDirectories="&quot;..\..\..\binaries\tcltk\lib&quot;;c:\tcl\lib"
```
remove &quot

kjakkali commented on win_configure/projects.VS2008/gen_server_ecl.bat:
```diff
@@ -16,7 +16,7 @@ echo "delete file not a success"
 )
 echo "Parsing the Master file for server and ecl"
-"C:\Program Files\Python27\python.exe" "../../buildutils/attr_parser.py" -m "../../src/lib/Libattr/master_job_attr_def.xml" -s "../../src/lib/Libattr/job_attr_def.c" -e "../../src/lib/Libecl/ecl_job_attr_def.c" -a job
+"..\..\..\binaries\python\python.exe" "../../buildutils/attr_parser.py" -m "../../src/lib/Libattr/master_job_attr_def.xml" -s "../../src/lib/Libattr/job_attr_def.c" -e "../../src/lib/Libecl/ecl_job_attr_def.c" -a job
```
The exe path follows windows style where as other parameters follow unix style. Please make it uniform.

hirenvadalia commented:
@kjakkali Thanks for review I have addressed your comments, please have a look into latest commit and let me know your thoughts.

hirenvadalia commented:
Also I have re-based branch with latest master as it was showing conflicts in some of files.

hirenvadalia commented:
Please check [this](https://ci.appveyor.com/project/hirenvadalia/pbspro/build/%23232) for latest build report on AppVeyor.

hirenvadalia commented:
Hey @kjakkali I have addressed your comments, please have a look into latest commit and let me know your thoughts.

Please check [this](https://ci.appveyor.com/project/hirenvadalia/pbspro/build/%23233) for latest AppVeyor build report.

hirenvadalia commented:
@kjakkali and @arungrover Thank you guys for review. @arungrover I have re-based branch with latest master and this PR is ready for merge. So, please take it further. 😄

subhasisb commented:
Merged it before it goes out of sync again. Thanks all.

hirenvadalia commented:
@subhasisb Thanks a lot for merging.
